Resiliency docs (#2326)
dapr/docs: daprdocs/content/en/operations/resiliency/resiliency-overview.md

---
type: docs
title: "Overview"
linkTitle: "Overview"
weight: 4500
description: "Configure Dapr retries, timeouts, and circuit breakers"
---
{{% alert title="Note" color="primary" %}}
 Resiliency is currently a preview feature. Before you can utilize a resiliency spec, you must first [enable the resiliency preview feature]({{< ref support-preview-features >}}).
{{% /alert %}}

> Distributed applications are commonly comprised of many moving pieces, with dozens, even hundreds, of instances for any given service. With so many moving pieces, the likelihood of a system failure increases. For example, an instance can fail due to hardware failures, an overwhelming number of requests, application restarts/scale outs, or any other reason. These events can cause a network call between services to fail. Designing and implementing your application with fault tolerance (the ability to detect, mitigate, and respond to failures) allows your application to recover quickly to a functioning state.

Reviewer: ```suggestion
Distributed applications are commonly comprised of many microservices, with dozens, even hundreds, of instances for any given application. With so many microservices, the likelihood of a system failure increases. For example, an instance can fail or be unresponsive due to hardware, an overwhelming number of requests, application restarts/scale outs, or several other reasons. These events can cause a network call between services to fail. Designing and implementing your application with fault tolerance, the ability to detect, mitigate, and respond to failures, allows your application to recover to a functioning state and become self healing.
```

> Dapr provides a mechanism for defining and applying fault tolerance/resiliency policies via a [resiliency spec]({{< ref "resiliency-overview.md#complete-example-policy" >}}). The resiliency spec is defined in the same location as components and is applied when the Dapr sidecar starts.  The sidecar determines when and how to apply resiliency policies to your Dapr API calls. In self-hosted mode, the resiliency spec must be named `resiliency.yaml`. In Kubernetes Dapr scans all resiliency specs. Within the resiliency spec, you can define policies for popular resiliency patterns, such as:

Reviewer: ```suggestion
Dapr provides a capability for defining and applying fault tolerance resiliency policies via a [resiliency spec]({{< ref "resiliency-overview.md#complete-example-policy" >}}). Resiliency specs are saved in the same location as components specs and are applied when the Dapr sidecar starts.  The sidecar determines how to apply resiliency policies to your Dapr API calls. In self-hosted mode, the resiliency spec must be named `resiliency.yaml`. In Kubernetes Dapr finds the named resiliency specs used by your application. Within the resiliency spec, you can define policies for popular resiliency patterns, such as:
```

Reviewer: I also raised this issue https://github.com/dapr/cli/issues/953 to rename --components-path to a better name.

- [Timeouts]({{< ref "policies.md#timeouts" >}})
- [Retries/back-offs]({{< ref "policies.md#retries" >}})
- [Circuit breakers]({{< ref "policies.md#circuit-breakers" >}})

Policies can then be applied to [targets]({{< ref "targets.md" >}}), which include:

- [Apps]({{< ref "targets.md#apps" >}}) via service invocation
- [Components]({{< ref "targets.md#components" >}})
- [Actors]({{< ref "targets.md#actors" >}})

Additionally, resiliency policies can be [scoped to specific apps]({{< ref "component-scopes.md#application-access-to-components-with-scopes" >}}).

Below is the general structure of a resiliency policy:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: myresiliency
scopes:
  # optionally scope the policy to specific apps
spec:
  policies:
    timeouts:
      # timeout policy definitions

    retries:
      # retry policy definitions

    circuitBreakers:
      # circuit breaker policy definitions

  targets:
    apps:
      # apps and their applied policies here

    actors:
      # actor types and their applied policies here

    components:
      # components and their applied policies here
```

> ### Complete Example Policy

Reviewer: ```suggestion
### Complete example policy
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: myresiliency

> # simialrly to Subscriptions and Configurations specs, scopes lists the Dapr App IDs that this

Reviewer: ```suggestion
# similar to subscription and configuration specs, scopes lists the Dapr App IDs that this
```

> # configuration applies to.

Reviewer: ```suggestion
# resiliency spec can be used by.
```

scopes:
  - app1
  - app2
spec:
  # policies is where timeouts, retries and circuit breaker policies are defined. 
  # each is given a name so they can be referred to from the targets section in the resiliency spec.
  policies:
    # timeouts are simple named durations.
    timeouts:
      general: 5s
      important: 60s
      largeResponse: 10s

    # retries are named templates for retry configurations and are instantiated for life of the operation.
    retries:
      pubsubRetry:
        policy: constant
        duration: 5s
        maxRetries: 10

      retryForever:
        policy: exponential
        maxInterval: 15s
        maxRetries: -1 # retry indefinitely

      important:
        policy: constant
        duration: 5s
        maxRetries: 30

      someOperation:
        policy: exponential
        maxInterval: 15s

      largeResponse:
        policy: constant
        duration: 5s
        maxRetries: 3

>     # circuit breakers are automatically instantiated per component and app endpoint.

Reviewer: Should this be "per component instance and app instance endpoint"?

>     # circuit breakers maintain counters that can live as long as the Dapr sidecar.

Reviewer: ```suggestion
    # circuit breakers maintain counters that live as long as the Dapr sidecar is running. They are not persisted.
```

    circuitBreakers:
      simpleCB:
        maxRequests: 1
        timeout: 30s 
        trip: consecutiveFailures >= 5

      pubsubCB:
        maxRequests: 1
        interval: 8s
        timeout: 45s
        trip: consecutiveFailures > 8

  # targets are what named policies are applied to. Dapr supports 3 target types - apps, components and actors
  targets:
    apps:
      appB:
        timeout: general
        retry: important

>         # circuit breakers for services are scoped per endpoint (e.g. hostname + port).

Reviewer: this e.g. is wrong is seems.  And endpoint is a method name which an app can have many off. If a circuit breaker applies to host + port then it applies to all endpoints (method) on a service. Is this correct? So the question is at what level do CB apply, to a single method or all method on an service?

        # when a breaker is tripped, that route is removed from load balancing for the configured `timeout` duration.
        circuitBreaker: simpleCB

    actors:
      myActorType: # custom Actor Type Name
        timeout: general
        retry: important

>         # circuit breakers for actors are scoped by type, id, or both.

Reviewer: How do I specify the difference between applying this to the actor type or actor id?

        # when a breaker is tripped, that type or id is removed from the placement table for the configured `timeout` duration.
        circuitBreaker: simpleCB

>         circuitBreakerScope: both ## 

Reviewer: What are the values I can have here? "type", "id", "both"?

        circuitBreakerCacheSize: 5000

    components:
      # for state stores, policies apply to saving and retrieving state.
      statestore1: # any component name -- happens to be a state store here
        outbound:
          timeout: general
          retry: retryForever

>           # circuit breakers for components are scoped per component configuration/instance (e.g. redis1).

Reviewer: ```suggestion
          # circuit breakers for components are scoped per component configuration/instance. For example myRediscomponent.
```

          # when this breaker is tripped, all interaction to that component is prevented for the configured `timeout` duration.
          circuitBreaker: simpleCB

      pubsub1: # any component name -- happens to be a pubsub broker here
        outbound:
          retry: pubsubRetry
          circuitBreaker: pubsubCB

      pubsub2: # any component name -- happens to be another pubsub broker here
        outbound:
          retry: pubsubRetry
          circuitBreaker: pubsubCB

>         inbound: # inbound only applies to delivery from sidecar to app

Reviewer: I do not understand this. There is also a timeout on the DB (line 118 above) which may get confusing with the timeout:general. How do these work together and can we given an example?

---
type: docs
title: "Overview"
linkTitle: "Overview"
weight: 4500

> description: "Configure Dapr error retries, timeouts, and circuit breakers"

Reviewer: Are the words "Dapr error" needed in this title? How about "Configure retries, timeouts, and circuit breakers"

Author: Done.

---

> > Resiliency is currently a preview feature. Before you can utilize resiliency policies, you must first [enable the resiliency preview feature]({{< ref preview-features >}}).

Reviewer: I would give the configuration needed to enable this so that people can cut an paste, or point to the article with the setting. This one http://localhost:1313/operations/support/support-preview-features/

## Introduction

Distributed applications are commonly comprised of many moving pieces, with dozens, even hundreds, of instances for any given service. With so many moving pieces, the likelihood of a system failure increases. For example, an instance can fail due to hardware failures, an overwhelming number of requests, application restarts/scale outs, or any other reason. These events can cause a network call between services to fail. Designing and implementing your application with fault tolerance (the ability to detect, mitigate, and respond to failures) allows your application to recover quickly to a functioning state.

## Overview

> Dapr provides a mechanism for defining and applying fault tolerance/resiliency policies via a [resiliency spec]({{< ref "resiliency-overview.md#complete-example-policy" >}}). The resiliency spec is defined in the same location as components and is applied when the Dapr sidecar starts. The sidecar determines when and how to apply resiliency policies to your Dapr API calls. Within the resiliency spec, you define policies for popular resiliency patterns, such as:

Reviewer: ```suggestion
Dapr provides a mechanism for defining and applying fault tolerance/resiliency policies via a [resiliency spec]({{< ref "resiliency-overview.md#complete-example-policy" >}}). The resiliency spec is defined in the same location as components and is applied when the Dapr sidecar starts. The sidecar determines when and how to apply resiliency policies to your Dapr API calls. Within the resiliency spec, you can define policies for popular resiliency patterns, such as:
```

Reviewer: One thing that is missing is a CLI command to list the available resiliencies policies. See this issue https://github.com/dapr/cli/issues/950.

- [Timeouts]({{< ref "policies.md#timeouts" >}})
- [Retries/back-offs]({{< ref "policies.md#retries" >}})
- [Circuit breakers]({{< ref "policies.md#circuit-breakers" >}})

> Policies can then be applied consistently to [targets]({{< ref "targets.md" >}}), which include:

Reviewer: ```suggestion
Policies can then be applied to [targets]({{< ref "targets.md" >}}), which include:
```

- [Apps]({{< ref "targets.md#apps" >}}) via service invocation
- [Components]({{< ref "targets.md#components" >}})
- [Actors]({{< ref "targets.md#actors" >}})

Additionally, resiliency policies can be [scoped to specific apps]({{< ref "component-scopes.md#application-access-to-components-with-scopes" >}}).

> Below is the general structure of what a resiliency policy looks like:

Reviewer: ```suggestion
Below is the general structure of a resiliency policy:
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:

>   name: resiliency

Reviewer: Suggest giving this a name that is different from the kind, just to make this clear that you can list all of these policies by name
```suggestion
  name: myresiliency
```

scopes:
  # optionally scope the policy to specific apps
spec:
  policies:
    timeouts:
      # timeout policy definitions

    retries:
      # retry policy definitions

    circuitBreakers:
      # circuit breaker policy definitions

  targets:
    apps:
      # apps and their applied policies here

    actors:
      # actor types and their applied policies here

    components:
      # components and their applied policies here
```

> > Note: In self-hosted mode, the resiliency policy must be named `resiliency.yaml` and reside in the components folder provided to the sidecar. In Kubernetes, Dapr scans all resiliency policies.

Reviewer: ```suggestion
> Note: In self-hosted mode, the resiliency policy must be named `resiliency.yaml` and reside in the components folder provided to the sidecar. In Kubernetes, resiliency policies are global and a named policy if loaded by the applications's sidecar .
```

Reviewer: "Scans all resiliency policies" is ambiguous. This come back to the need for a CLI command and also the ability to see these policies in the CLI and UI. Really this is the ability to have "named" policies that you can reference from you application.

Author: @msfussell I don't follow this. I think what we're actually saying here is that  "resiliency specs are global"?

> > Note: In self-hosted mode, the resiliency policy must be named `resiliency.yaml` and reside in the components folder provided to the sidecar. In Kubernetes, Dapr scans all resiliency policies.

Reviewer: I do not think this should be a note. This should be front and center. In. K8s you can deploy multiple named policies. In self-hosted you can only have one and it must be a file called resiliency.yaml. These should be heading in this document for each environment.

Author: @msfussell I think we're confusing the term resiliency policies with resiliency specs. A policy is a named definition for one of the supported resiliency patterns. A spec is the grouping / definitions of named policies.

### Complete Example Policy

```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:

>   name: resiliency

Reviewer: Use a different name to make this clear that this is a named policy

> # simialrly to Subscriptions CRD, scopes lists the Dapr App IDs that this

Reviewer: This is a confusing comment. Configurations can also be scoped. And we should use the word "spec" and not CRD. Make this consistent. 

```suggestion
# simialrly to Subscriptions CRD, scopes lists the Dapr App IDs that this
```

    # circuit breakers are automatically instantiated per component and app endpoint.
    # circuit breakers maintain counters that can live as long as the Dapr sidecar.
    circuitBreakers:
      simpleCB:
        maxRequests: 1
        timeout: 30s 
        trip: consecutiveFailures >= 5

      pubsubCB:
        maxRequests: 1
        interval: 8s
        timeout: 45s
        trip: consecutiveFailures > 8

  # targets are what named policies are applied to. Dapr supports 3 target types - apps, components and actors
  targets:
    apps:
      appB:
        timeout: general
        retry: important
        # circuit breakers for services are scoped per endpoint (e.g. hostname + port).
        # when a breaker is tripped, that route is removed from load balancing for the configured `timeout` duration.
        circuitBreaker: simpleCB

    actors:
      myActorType: # custom Actor Type Name
        timeout: general
        retry: important
        # circuit breakers for actors are scoped by type, id, or both.
        # when a breaker is tripped, that type or id is removed from the placement table for the configured `timeout` duration.
        circuitBreaker: simpleCB

>         circuitBreakerScope: both

Reviewer: What is this "both"?

Author: This can be found in the circuit breaker section where we explain all the CB properties

>         circuitBreakerCacheSize: 5000

Reviewer: What is this cachsize? 5000 of what measurement specifically?

Author: Same as above - this can be found in the circuit breaker section where we explain all the CB properties
